### PR TITLE
Ensure git index and workdir match expectations when switching branches with sparse-checkout.

### PR DESCRIPTION
**Pull request by erikmchut**
## Overview

With a sparse-checkout repo, switching branches with git_checkout_tree has been resulting in an incorrect index entry following the action. Files that should not be checked out (ls-files status == 'S') keep their source blob id and status following the branch switch, leading to an index that appears to have modifications.

## Fix

This PR removes logic that was skipping diff generation for no-checkout files, as the diff needs to exist to be applied to the index. Instead, it adds logic to other parts of the code to ensure the index is correctly modified without adding no-checkout files to the working directory.

### CODE REVIEW

File: src/libgit2/checkout.c [hunk 1]
```diff
@@ -2578,7 +2586,6 @@ int git_checkout_iterator(
 		GIT_DIFF_INCLUDE_TYPECHANGE_TREES |
 		GIT_DIFF_SKIP_BINARY_CHECK |
 		GIT_DIFF_INCLUDE_CASECHANGE;
-	diff_opts.skip_sparse_files = 1;
```
Comment (cmbartschat) [inline src/libgit2/checkout.c:2581]: Is `skip_sparse_files` an option we added that we can delete?

File: src/libgit2/checkout.c [hunk 2]
```diff
@@ -516,9 +520,13 @@ static int checkout_action_with_wd(
 			*action = CHECKOUT_ACTION_IF(FORCE, UPDATE_BLOB, CONFLICT);
 		break;
 	case GIT_DELTA_DELETED: /* case 9 or 10 (or 26 but not really) */
-		if (checkout_is_workdir_modified(data, &delta->old_file, &delta->new_file, wd))
-			*action = CHECKOUT_ACTION_IF(FORCE, REMOVE, CONFLICT);
-		else
+		if (checkout_is_workdir_modified(data, &delta->old_file, &delta->new_file, wd)) {
+			if (git_iterator_current_skip_checkout(workdir)) {
```
Comment (cmbartschat) [inline src/libgit2/checkout.c:524]: It's possible this logic doesn't come into play - if you have modified changes to a workdir file that shouldn't even be there, maybe `CONFLICT` is the right thing.

File: src/libgit2/checkout.c (hunk 1 above)
Comment (erikmchut) [reply src/libgit2/checkout.c:2581]: I'll remove it as we're no longer using it following this PR.

File: src/libgit2/checkout.c (hunk 2 above)
Comment (erikmchut) [reply src/libgit2/checkout.c]: That appears to be the case. I've removed it.

**Review by cmbartschat — APPROVED** [review]

**Pull request merged**